File: backend/internal/services/groq/whisper.go
package groq

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"time"
)

type WhisperService struct {
	apiKey string
	model  string
	client *http.Client
}

type WhisperResponse struct {
	Text string `json:"text"`
}

// Optimized HTTP client with connection pooling for low latency
var optimizedTransport = &http.Transport{
	MaxIdleConns:        100,
	MaxIdleConnsPerHost: 10,
	IdleConnTimeout:     90 * time.Second,
	DisableCompression:  false,
	ForceAttemptHTTP2:   true,
}

func NewWhisperService(apiKey, model string) *WhisperService {
	return &WhisperService{
		apiKey: apiKey,
		model:  model,
		client: &http.Client{
			Transport: optimizedTransport,
			Timeout:   60 * time.Second, // 60s timeout for audio transcription
		},
	}
}

func (s *WhisperService) Transcribe(ctx context.Context, audioData []byte, language string) (string, error) {
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	// Add audio file
	part, err := writer.CreateFormFile("file", "audio.wav")
	if err != nil {
		return "", err
	}
	if _, err := part.Write(audioData); err != nil {
		return "", err
	}

	// Add model
	writer.WriteField("model", s.model)
	if language != "" {
		writer.WriteField("language", language)
	}
	writer.WriteField("response_format", "json")

	writer.Close()

	// Create request
	req, err := http.NewRequestWithContext(
		ctx,
		"POST",
		"https://api.groq.com/openai/v1/audio/transcriptions",
		body,
	)
	if err != nil {
		return "", err
	}

	req.Header.Set("Authorization", "Bearer "+s.apiKey)
	req.Header.Set("Content-Type", writer.FormDataContentType())

	// Send request
	resp, err := s.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("groq whisper API error (status %d): %s", resp.StatusCode, string(bodyBytes))
	}

	var whisperResp WhisperResponse
	if err := json.NewDecoder(resp.Body).Decode(&whisperResp); err != nil {
		return "", err
	}

	return whisperResp.Text, nil
}

File: desktop/src/services/groq/whisper.ts
import { configService } from '../api/config'

export interface WhisperResult {
  text: string
}

class WhisperService {
  private apiUrl = 'https://api.groq.com/openai/v1/audio/transcriptions'

  async transcribe(audioBlob: Blob, language = 'en'): Promise<string> {
    const config = await configService.getConfig()

    const formData = new FormData()
    formData.append('file', audioBlob, 'audio.webm')
    formData.append('model', config.groqWhisperModel || 'whisper-large-v3-turbo')
    formData.append('language', language)
    formData.append('response_format', 'json')

    const response = await fetch(this.apiUrl, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${config.groqApiKey}`,
      },
      body: formData,
    })

    if (!response.ok) {
      const error = await response.text()
      throw new Error(`Groq API error: ${response.status} - ${error}`)
    }

    const result: WhisperResult = await response.json()
    return result.text
  }
}

export const whisperService = new WhisperService()
